Make SMT Printer warning logs more pythonic.

kunalsheth opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
Fix https://github.com/sympy/sympy/pull/23961.


#### Brief description of what is fixed or changed
Made SMT printer tests not send logs to STDOUT and instead use python `warnings` module as noted here: https://github.com/sympy/sympy/pull/23961#issuecomment-1235844014


#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

kunalsheth committed:
Make warning logs more pythonic.

kunalsheth committed:
Fix bad import.

asmeurer commented on sympy/printing/smtlib.py:
```diff
@@ -248,23 +247,24 @@ def smtlib_code(
     Examples
     ========
+    >>> import warnings
+    >>> warnings.showwarning = (lambda m, *_: print(repr(m)))
```
Does doctest support warnings? Changing `showwarnings` like this is problematic because it changes the warnings module globally, meaning it will still be in effect after the doctests run.

kunalsheth replied:
It doesn't seem like there's a great way to do this: https://stackoverflow.com/questions/2418570/testing-warnings-with-doctest

If you think it's sufficient, I can just make sure to turn `showwarnings` back when I'm done using it?

asmeurer replied:
I guess this is fine for now. My guess is we can pipe stderr to stdout [here](https://github.com/sympy/sympy/blob/8a5d62234c0713fa75c236193295cb21afb3ef8d/sympy/testing/runtests.py#L1815) to make warnings work in doctests, but I haven't tried it so I can't say for sure how straightforward it will be.

oscarbenjamin commented:
Why are these warnings being issued instead of just errors?

kunalsheth committed:
Turn showwarning lambda back to original after doctest.

oscarbenjamin commented on sympy/printing/smtlib.py:
```diff
@@ -338,7 +338,7 @@ def smtlib_code(
             ):
-                log_warn(f"Could not infer type of `{sym}`. Defaulting to float.")
+                warnings.warn(f"Could not infer type of `{sym}`. Defaulting to float.")
                 p.symbol_table[sym] = float
```
Why not just raise an error here?

kunalsheth replied:
Because I don't think these qualify as program-crashing errors. I speculate that a large majority of sympy queries are going to be over real numbers, so it would be excessively pedantic and cumbersome to require users to always specify that.

I am trying to serve 2 groups of (imagined) users:
1. Many users just want a more powerful solver that can handle constraints. They don't want to have to deal with the specifics of SMT and don't want to have to write-out any more information to the SMT solver than would be required for a regular sympy `solve()` call.
2. There are also more advanced users, specifically from academia, who care deeply about the specifics of what the solver is acting on. These users care about being precise because they are trying to construct robust mathematical _proofs_ (not just trying to solve an equation). They will likely make full use of `symbol_table`, `known_functions`, etc. and would want a very explicit API which warns loudly of any ambiguity in their query.

Making these errors would hurt group 1 and making these silent would hurt group 2. Thus I decide that warnings are the best way forward.

oscarbenjamin replied:
Issuing warnings when something is used in the way that it is expected to be used is just bad interface design IMO.

Why not just document that if `symbol_table` is not passed then all symbols will be assumed to be floats (since I presume that's the behaviour here)? Then if `symbol_table` is passed but doesn't contain the symbol an exception can be raised.

sylee957 commented:
There are many warning logs printed out during the testing
https://github.com/sympy/sympy/actions/runs/3004717888/jobs/4824274707#step:6:511
but these could be error instead.
Can you resolve this?

kunalsheth committed:
Mute warnings in SMTLib print tests.

oscarbenjamin commented on sympy/printing/tests/test_smtlib.py:
```diff
@@ -13,30 +14,41 @@
 
 
 def test_Integer():
+    warning_filters = list(warnings.filters)
+    warnings.simplefilter('ignore')
     assert smtlib_code(Integer(67)) == "67"
     assert smtlib_code(Integer(-1)) == "-1"
+    warnings.filters = warning_filters
```
This kind of thing would need to be guarded with `try/finally` but in any case there is the `ignore_warnings` context manager for this:
https://github.com/sympy/sympy/blob/e8c22f6eac7314be8d92590bfff92ced79ee03e2/sympy/functions/elementary/tests/test_miscellaneous.py#L349-L350

asmeurer replied:
We can make a top-level helper in the test file for this that runs the smtlib_code function and catches the warnings, so you don't have to do this for every single test (maybe even have the function assert that certain warnings are emitted).

kunalsheth replied:
fixed!

kunalsheth committed:
Create warnings context manager for smtlib print test.

oscarbenjamin commented on sympy/printing/tests/test_smtlib.py:
```diff
@@ -7,131 +13,183 @@
+
+@contextlib.contextmanager
+def _smtlib_warns(expected: typing.Iterable[_W]):
+    # based on sympy.testing.pytest.warns()
```
Why isn't this just using the existing `warns` function?

oscarbenjamin commented:
There is a lot of discussion above and in the related issues about how exactly to handle warnings including how to suppress them in the test suite and so on but I think this all misses the point:

- An API that issues warnings when used correctly is a bad API.

This should be changed to have an API that has documented behaviour for valid inputs and gives errors for invalid input. That is always the best option when designing a new API. Every case where this code emits a warning should be changed to either

- Raise an exception, or...
- Follow a clearly documented behaviour and *not* emit a warning.

Adding warnings can make sense when there is a need to preserve compatibility with previous behaviour because we already have a badly designed API and the only reasonable way to continue to honour it is to do so with a warning.

It might make sense to reconsider the whole approach for how the functions work and what their signatures are. Ideally an API should be designed in such a way that it isn't even possible for someone to call it with arguments that would necessitate any warning but I don't know if that's possible here.

kunalsheth commented:
@oscarbenjamin  I understand where you are coming from— I am a big fan of type-safe and clean API design. I’d agree, if I were designing e.g. a REST API for a microservice, releasing warnings would not be desirable. APIs need to be more black-and-white like you said.

**I think the key misunderstanding here is calling this an “API.”** `smtlib_code(..)` is not an _Application_ Programming Interface, it is a _Tool_. It is meant to be called by humans (not machines) and likely from an interactive notebook or other scientific environment. Furthermore, the humans calling `smtlib_code(..)` are likely Formal Methods researchers who have different priorities from Software Engineers. 

**Cumbersome automated reasoning APIs already exist, but great _tools_ don't.**
Good tools have different design priorities than good APIs. Thus, I don't think this black-and-white thinking of “good API” v.s. “bad API” is particularly constructive here, and applying it so broadly won’t produce a genuinely helpful product. The design of this tool needs to be nuanced, ergonomic, and involves domain specific knowledge. To recap the targeted users of this tool: https://github.com/sympy/sympy/pull/24009#discussion_r965199910

## We are not building an _API_ because SMT-backed _Applications_ would not, or cannot, use it:
- SMT solvers already have their own, far more mature, APIs. Creating yet-another-SMT-API doesn't have much value. See: [#1](https://pypi.org/project/z3-solver/) [#2](https://github.com/pysmt/pysmt) [#3](https://pypi.org/project/dreal/) [#4](https://pypi.org/project/cvc5/).
- SMT-backed _Applications_ tend to operate by parsing PLs/DSLs and forming queries based on an AST or other representation irrelevant to Sympy.
- SMT-backed Formal Verification _Applications_ have the most stringent soundness requirements (generating rigorous proofs, verifying said proof, verifying the verifier of said proof, verifying the verifier of the verifier, etc). Sympy does not meet this level of assurance and is not an acceptable dependency to call into.

## Addressing some of your comments:
> Adding warnings can make sense when there is a need to preserve compatibility with previous behaviour because we already have a badly designed API and the only reasonable way to continue to honour it is to do so with a warning.

Yes. The "badly designed" (or rather, just a bit ambiguous) API I am trying to "preserve compatibility with" is Sympy. SMT-Lib is more precise and more robust, particularly when it comes to the types of variables. This is why I need to emit warnings when I encounter potentially ambiguous queries generated by Sympy. For example, Try and figure out what `Symbol('x')` is when `sympy.Symbol('x').is_integer is None` and `sympy.Symbol('x').is_real is None` and every permutation of `sympy.Symbol('x', bool=True).is_Boolean` is False. The information is likely there, but might be implied. Or maybe it isn't, but a guess can be drawn from context.


> Issuing warnings when something is used in the way that it is expected to be used is just bad interface design IMO.

An auditable log of all assumptions made is very desirable in Formal Methods. SMT solvers themselves do this in the form of step-by-step proofs. These proofs can then be verified by additional tooling.


> Why not just document that if symbol_table is not passed then all symbols will be assumed to be floats (since I presume that's the behaviour here)? Then if symbol_table is passed but doesn't contain the symbol an exception can be raised.

That's not the behavior here, it is more complex than that. I also don't do that because the user experience would be awful, especially when queries get slightly larger. Real numbers actually see very little support amongst SMT solvers unlike integers, uninterpreted functions, and booleans. Sympy, on the other hand, _almost_ defaults to real numbers, but it's good to log this assumption.


> This should be changed to have an API that has documented behaviour for valid inputs and gives errors for invalid input. That is always the best option when designing a new API. Every case where this code emits a warning should be changed to ... Follow a clearly documented behaviour and not emit a warning.

I think this philosophy might have some value for APIs, but again, I'm making a tool. Being pedantic about “following clearly documented behavior” doesn’t really help human beings. For example, the Python interpreter has a “clearly documented” specification, so why would anyone use PyCharm IDE instead of plain TextEdit? 


> “An API that issues warnings when used correctly is a bad API.”

A tool that operates silently in ambiguous cases is a bad tool, especially when the point of said tool is to verify the correctness of something. For example, a C compiler that always lets undefined behavior pass silently would be excruciating to use when debugging. Yet, a C compiler that never lets undefined behavior pass would be near useless for embedded systems applications. The middle ground is compiler warnings. It’s all about balance, which is why I don’t think this black-and-white thinking of “good API” versus “bad API” is particularly constructive.

oscarbenjamin commented:
Feel free to replace the word "API" with "interface" or whatever you want. That does not change the point.

Can you show a clear example of why you think warnings need to be given so that we can discuss that?

kunalsheth commented:
@oscarbenjamin @asmeurer  I don't feel I'm being heard. I've written a nuanced and comprehensive explanation which I feel was just dismissed without any consideration.

If the issue here is specifically warnings themselves, I propose 1 way we can move forward is to revert back to my original `log_warn: lambda function, optional` with default value `(lambda _: None)`. This way, no warnings/logs will be emitted unless explicitly requested by users.

oscarbenjamin commented:
Okay, here is a concrete example:
```python
In [16]: smtlib_code(x < 0)
Could not infer type of `x`. Defaulting to float.
Out[16]: '(declare-const x Real)\n(assert (< x 0))'
```
Is this really ambiguous? Why can't this default to float without printing a warning? It isn't possible here for `x` to be a boolean. It could be an `int` rather than a `float` but why is it bad to just default to `float` here?

In this example `x` is implicitly treated as a `bool` with no warning:
```python
In [17]: smtlib_code(x)
Out[17]: '(declare-const x Bool)\n(assert x)'
```
What exactly is the difference between these cases such that one needs a warning and the other does not?

oscarbenjamin commented:
> If the issue here is specifically warnings themselves

The warnings are one thing: Python generally eschews warnings. Your example of a C compiler is an interesting choice because Python as a language has since its inception deliberately avoided emulating the poor UX of C's compiler warnings in favour of showing single clear exceptions one at a time. It is not uncommon that someone will want to add warnings to some part of SymPy but unless it's a deprecation warning it almost always turns out to be a mistake that causes more problems than it solves.

I'm trying to understand why the warnings are there at all though because it suggests that the interface is ambiguous and is therefore something that we would end up regretting in future.

oscarbenjamin commented:
Another possibility by the way is to have an optional argument that prints information to stdout like `smtlib_code(expr, fulloutput=True)`. Then if someone wants to see what assumptions are being made they can. If it's optional (and off by default) then it's not a problem to print to stdout. It's still better to return that full output rather than print it directly to stdout though like `code, info = smtlib_code(expr, fulloutput=True)`: then the user can control where and when to display `info`.

oscarbenjamin commented:
Something like this makes me wonder if either there should be a separate function for printing expressions (rather than booleans) or if it should just be disallowed:
```python
In [18]: smtlib_code(x**2 - y)
Could not infer type of `y`. Defaulting to float.
Could not infer type of `x`. Defaulting to float.
Non-Boolean expression `x**2 - y` will not be asserted. Converting to SMTLib verbatim.
Out[18]: '(declare-const x Real)\n(declare-const y Real)\n(+ (pow x 2) (* -1 y))'
```
Is this even valid smtlib code? My understanding is that in smtlib everything should be a boolean statement.

sylee957 commented:
What about separating the automated inference as an user API, 
and give option for users to review the inference result instead of making users to review the warnings?

https://github.com/sympy/sympy/blob/5ba43252efd64bcda5aa0f6bc27debdb6b0fefc2/sympy/printing/smtlib.py#L308-L311

kunalsheth committed:
Undo #24009.

kunalsheth committed:
Fix style error on untyped list.

oscarbenjamin commented:
Thanks. I think this looks good. The new behaviour is that there is an optional argument that can be used to print warnings to stdout but the output is off by default. I think that's a better UX.

I'll merge this now.

oscarbenjamin commented:
The three warnings are:
```
    DEFAULTING_TO_FLOAT = re.compile("Could not infer type of `.+`. Defaulting to float.", re.I)
    WILL_NOT_DECLARE = re.compile("Non-Symbol/Function `.+` will not be declared.", re.I)
    WILL_NOT_ASSERT = re.compile("Non-Boolean expression `.+` will not be asserted. Converting to 
```
If I understand correctly the first warning relates to the fact that SymPy allows an ordinary Symbol to be both a Boolean and an Expr simultaneously. In context we can sometimes infer the meaning e.g. `x + y` means that `x` and `y` are `Expr` and `x & y` means that they are `Boolean` but if we just have `x` then we don't know which applies.

I agree that this is a problem and I have been saying for a long time that there should be a separate `BooleanSymbol` type that should be used for symbols representing boolean values. As far as I know Symbol is the only type in SymPy that is both Boolean and Expr. There are various places in the codebase that have awkward workarounds for the fact that it isn't straight-forward to distinguish boolean symbols cleanly.

The kind system should be a better way to distinguish these things and we made sure that Symbol.kind always gives NumberKind but there needs to be a way to make an explicit boolean symbol so that it can have kind BooleanKind.

The other two warnings types seem to be about allowing users to print a fragment of smtlib code rather than a complete program if I understand correctly. I would probably make it two separate user functions where one is for fragments and the other is for complete programs.

kunalsheth commented:
> The other two warnings types seem to be about allowing users to print a fragment of smtlib code rather than a complete program if I understand correctly. I would probably make it two separate user functions where one is for fragments and the other is for complete programs.

`WILL_NOT_ASSERT` is meant to be used for solver-specific builtins. There are many more commands than just "assert" in SMT-Lib, and many solvers have their own non-standard commands too. Here is an example for `WILL_NOT_ASSERT` occurring "correctly" in a complete program.

Notebook:
```python
x, y, z = symbols('x y z', integer=True)
minimize = Function('minimize')  # special solver-specific builtin
known_functions = SMTLibPrinter()._known_functions | {minimize: minimize.name}
print(smtlib_code(
    [
        (-1 <= x) & (x < 10),
        y < z,
        x ** 2 + y ** 2 + z ** 2 > 20,
        minimize(z)
    ],
    known_functions=known_functions,
    log_warn=print
))
```
Output:
```
Non-Boolean expression `minimize(z)` will not be asserted. Converting to SMTLib verbatim.
(declare-const x Int)
(declare-const y Int)
(declare-const z Int)
(assert (and (>= x -1) (< x 10)))
(assert (< y z))
(assert (> (+ (pow x 2) (pow y 2) (pow z 2)) 20))
(minimize z)
```

oscarbenjamin commented:
Would it not be better to have something like:
```python
from sympy.printing.smtlib import Assert
print(smtlib_code(
    [
        Assert((-1 <= x) & (x < 10)),
        Assert(y < z),
        Assert(x ** 2 + y ** 2 + z ** 2 > 20),
        minimize(z)
    ],
))
```
Then the user specifies which things are supposed to be asserted or not and the printer doesn't need to try and guess that. Asserting is a key feature of smtlib so it seems strange to try and make it implicit.
